src/main/java/io/aiven/kafka/connect/gcs/TopicPartitionRecordGrouper.java

```diff
@@ -42,41 +45,54 @@
  */
 final class TopicPartitionRecordGrouper implements RecordGrouper {
-    private static final List<String> EXPECTED_VARIABLE_LIST = Arrays.asList(
-        FilenameTemplateVariable.TOPIC.name,
-        FilenameTemplateVariable.PARTITION.name,
-        FilenameTemplateVariable.START_OFFSET.name
-    );
 
     private final Template filenameTemplate;
 
     private final Integer maxRecordsPerFile;
 
+    private final TimestampSource tsSource;
+
     private final Map<TopicPartition, SinkRecord> currentHeadRecords = new HashMap<>();
 
-    private final Map<String, List<SinkRecord>> fileBuffers = new HashMap<>();
+    private final Map<RecordKey, List<SinkRecord>> fileBuffers = new HashMap<>();
+
+    private final Function<Parameter, String> setTimestamp;
 
     /**
      * A constructor.
      *
      * @param filenameTemplate  the filename template.
      * @param maxRecordsPerFile the maximum number of records per file ({@code null} for unlimited).
      */
-    public TopicPartitionRecordGrouper(final Template filenameTemplate, final Integer maxRecordsPerFile) {
+    public TopicPartitionRecordGrouper(final Template filenameTemplate,
+                                       final Integer maxRecordsPerFile,
+                                       final TimestampSource tsSource) {
         Objects.requireNonNull(filenameTemplate, "filenameTemplate cannot be null");
-
-        if (!acceptsTemplate(filenameTemplate)) {
-            throw new IllegalArgumentException(
-                "filenameTemplate must have set of variables {"
-                    + String.join(",", EXPECTED_VARIABLE_LIST)
-                    + "}, but {"
-                    + String.join(",", filenameTemplate.variables())
-                    + "} was given"
-            );
-        }
-
+        Objects.requireNonNull(tsSource, "tsSource cannot be null");
         this.filenameTemplate = filenameTemplate;
         this.maxRecordsPerFile = maxRecordsPerFile;
+        this.tsSource = tsSource;
+        this.setTimestamp = new Function<Parameter, String>() {
+
+            private ZonedDateTime lastTimestamp = tsSource.time();
+
+            @Override
+            public String apply(final Parameter parameter) {
+                final ZonedDateTime currentTime = tsSource.time();
+                if (parameter.value().equals("HH")
+                    && currentTime.getHour() > lastTimestamp.getHour()) {
+                    lastTimestamp = currentTime;
+                    return lastTimestamp.format(DateTimeFormatter.ofPattern(parameter.value()));
+                } else if (parameter.value().equals("dd")
+                    && currentTime.getDayOfMonth() > lastTimestamp.getDayOfMonth()) {
+                    lastTimestamp = currentTime;
+                    return lastTimestamp.format(DateTimeFormatter.ofPattern(parameter.value()));
+                } else {
+                    return currentTime.format(DateTimeFormatter.ofPattern(parameter.value()));
```

Comment:
I don't have a measurement, but looking at the code suggests `DateTimeFormatter` creation might not be a fast operation to perform on every record. I'd propose to create them once and put in some map with supported time units as keys.

Reply:
true. will fix